pkg/shipper/shipper.go

```diff
@@ -134,70 +137,195 @@ func (s *Shipper) Timestamps() (minTime, maxSyncTime int64, err error) {
+			return nil
+		}
+
+		// We only ship of the first compacted block level. See https://github.com/improbable-eng/thanos/issues/206
+		// for details. This is however not needed for continous Thanos performance. See standalone `sync` command
+		// on how to upload old, locally compacted blocks manually.
+		if m.Compaction.Level > 1 {
+			return nil
+		}
+
+		if m.Stats.NumSamples == 0 {
+			// Ignore empty blocks.
+			level.Debug(s.logger).Log("msg", "ignoring empty block", "block", m.ULID)
+			return nil
+		}
+
+		// Check against bucket if the meta file for this block exists.
+		ok, err := s.bucket.Exists(ctx, path.Join(m.ULID.String(), block.MetaFilename))
+		if err != nil {
+			return errors.Wrap(err, "check exists")
+		}
+		if ok {
+			return nil
+		}
+
+		if err := s.upload(ctx, m); err != nil {
+			level.Error(s.logger).Log("msg", "shipping failed", "block", m.ULID, "err", err)
+			// No error returned, just log line. This is because we want other blocks to be uploaded even
+			// though this one failed. It will be retried on second Sync iteration.
+			s.metrics.uploadFailures.Inc()
+			uploadErrs++
+			return nil
 		}
 		meta.Uploaded = append(meta.Uploaded, m.ULID)
+
+		uploaded++
+		s.metrics.uploads.Inc()
 		return nil
 	}); err != nil {
-		level.Error(s.logger).Log("msg", "iter block metas failed", "err", err)
-		return
+		s.metrics.dirSyncFailures.Inc()
+		return uploaded, errors.Wrap(err, "iter local block metas")
 	}
+
 	if err := WriteMetaFile(s.logger, s.dir, meta); err != nil {
 		level.Warn(s.logger).Log("msg", "updating meta file failed", "err", err)
 	}
-}
 
-func (s *Shipper) sync(ctx context.Context, meta *block.Meta) (err error) {
-	dir := filepath.Join(s.dir, meta.ULID.String())
+	s.metrics.dirSyncs.Inc()
+
+	if uploadErrs > 0 {
+		return uploaded, errors.Errorf("failed to sync %v blocks", uploadErrs)
 
-	// We only ship of the first compacted block level.
-	// TODO(bplotka): https://github.com/improbable-eng/thanos/issues/206
-	if meta.Compaction.Level > 1 {
-		return nil
 	}
+	return uploaded, nil
+}
 
-	// Check against bucket if the meta file for this block exists.
-	ok, err := s.bucket.Exists(ctx, path.Join(meta.ULID.String(), block.MetaFilename))
+// SyncAll performs a single synchronization for all compacted and non-compacted blocks. This ensures all compacted local blocks
+// that *can be* uploaded have been uploaded to the object bucket once. This function is *not* designed to be run
+// periodically. It should be used as a manual command to export old compacted blocks for migration purposes.
+//
+// NOTE: It is currently required to turn off compactor first to use this command safely. Otherwise non-fixable overlaps might happen.
+//
+// It is not concurrency-safe, however it is safe to run it with `shipper.Sync` in different binary.
+func SyncAll(
+	ctx context.Context,
+	logger log.Logger,
+	dir string,
+	bucket objstore.Bucket,
+	lbls func() labels.Labels,
+	source block.SourceType,
```

Comment:
Just thinking.. should be the SourceType preserved or possibly add new one `bucket.synced` to override the former?

Reply:
We have special bucket.sync for this (: so your idea is already there.